benches/comparative.rs

```diff
@@ -0,0 +1,64 @@
+use criterion::{black_box, criterion_group, criterion_main, Criterion};
+use rand::prelude::*;
+use std::collections::HashSet;
+
+const SIZE: usize = 10000;
+
+/// This benchmark compares the performance of HashSet, Vob, and Vec
+/// in a scenario where an algorithm needs to keep track of visited indices.
+#[cfg(test)]
+fn comparative(c: &mut Criterion) {
+    // Generate 10000 random (seeded) points
+    let mut rng: StdRng = SeedableRng::from_seed([42; 32]);
+    let mut indices:Vec<usize> = (0..SIZE).collect();
+
+    indices.shuffle(&mut rng);
+
+    c.bench_function("HashSet comparative", |b| {
+        b.iter(|| {
+            let mut hash_set = HashSet::new();
+            // set half of the set to "true"
+            for index in indices.iter().skip(SIZE / 2) {
+                black_box(hash_set.insert(index));
+            }
+
+            for index in &indices {
+                black_box(hash_set.contains(index));
```

Comment:
this usage of `black_box` only makes sure contains is used, both `index` and `hash_set` are still subject to optimization. 

This seems fine in this example, but I just wanted to flag it.

https://doc.rust-lang.org/std/hint/fn.black_box.html

Reply:
I think the objective is not to prevent optimization but to prevent the optimizer from removing the call altogether. 
If the compiler sees a side-effect free method call where the output is not used, it could elect not to do the call at all (that is at least my 5 cent understanding). 

Maybe we could accomplish the same thing by moving the black box outside the call?:

```rust
for index in &indices {
     let visited = hash_set.contains(index);
     black_box(!visited);
```

Reply:
Yeah, I think the main thing is that the compiler might still e.g. unroll this loop or const-eval `hash_set.contains()`.

Reply:
At some point we can become unduly paranoid. But I do see that you both have made a good point. Perhaps we need two levels:

```rust
for index in blackbox(&indices) {
  hash_set.contains(blackbox(index));
}
```

The first `blackbox` (hopefully) prevents the compiler realising that the `for` loop's length can be statically known. The second stops it realising that the index is (by defintion) contained in the loop. *But* since the `contains` has no side-effect, technically it can optimise it away. So I'd probably actually force a side effect:


```rust
for index in blackbox(&indices) {
  if !hash_set.contains(blackbox(index)) {
    panic!("...");
  }
}
```

Reply:
The latter should probably be written using `assert!()` instead of an explicit `panic!()`, though.

Reply:
The tests are setup so that only half of the values are "true", or else the Vob would get an unfair advantage by testing if the entire u32 is set. (not that i actually seen how that test is actually implemented).

We could re-arrange the setup so we could cross check if the bit should be true or not, though.

Reply:
Either way I think we're over-engineering this. @eadf could probably quickly check if additional `black_box` around index makes any difference and then put this to bed?